File: metaspace/output/exercises/solutions/04_sign_and_recover.ts
/**
 * Level 4 — Signing off-chain, verifying on the server.
 *
 * Use case:  A player connects a wallet to your game. There is no password to
 *            check and no account in your users table yet. The only thing you
 *            can ask for is a signature, and the only thing you can do with it
 *            is recover the address that produced it.
 * Purpose:   Understand that verification is *recovery* — you do not compare a
 *            signature to a stored secret, you reconstruct the signer from it.
 * Key facts: EIP-191 defines the "\x19Ethereum Signed Message:\n" prefix that
 *            makes a personal-signed message structurally unable to be a valid
 *            transaction. A signature is (r, s, v): 32 + 32 + 1 bytes.
 *
 * Run: npm run l4        (no network; throwaway key; nothing is sent anywhere)
 */

import {
  hashMessage,
  recoverMessageAddress,
  verifyMessage,
  keccak256,
  stringToHex,
  concatHex,
  slice,
  type Hex,
} from "viem";
import { privateKeyToAccount } from "viem/accounts";

// Anvil account #0 again: published in every toolchain's documentation, holds
// nothing, and exists so that examples like this need no secrets.
const player = privateKeyToAccount(
  "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80",
);

// ---------------------------------------------------------------------------
// Part 1 — sign, then recover.
// ---------------------------------------------------------------------------

console.log("--- Part 1: the round trip ---");

const message = "Log me in to Metaspace";
const signature = await player.signMessage({ message });

console.log("signer address :", player.address);
console.log("signature      :", signature);
console.log("  r :", slice(signature, 0, 32));
console.log("  s :", slice(signature, 32, 64));
console.log("  v :", slice(signature, 64, 65));

const recovered = await recoverMessageAddress({ message, signature });
console.log("recovered      :", recovered);
console.log("matches signer :", recovered === player.address);

console.log(
  "\nNothing was looked up. The server holds no key, no public key, and no prior\n" +
    "record of this player, and it still learned the address with certainty. That\n" +
    "is the property ECDSA gives you: the signature plus the message is enough to\n" +
    "reconstruct the public key, and the address is a hash of that public key.\n" +
    "verifyMessage() is the same operation with the comparison folded in:",
);

console.log(
  "verifyMessage  :",
  await verifyMessage({ address: player.address, message, signature }),
);

// ---------------------------------------------------------------------------
// Part 2 — the prefix, and why it exists.
// ---------------------------------------------------------------------------

console.log("\n--- Part 2: what actually got signed ---");

const prefix = stringToHex(`\x19Ethereum Signed Message:\n${message.length}`);
const byHand = keccak256(concatHex([prefix, stringToHex(message)]));

console.log("hashMessage(message) :", hashMessage(message));
console.log("built by hand        :", byHand);
console.log("identical            :", byHand === hashMessage(message));

console.log(
  "\nThe key never signed your message. It signed a hash of your message with a\n" +
    "fixed prefix glued to the front, and that prefix is a safety interlock. A\n" +
    "raw Ethereum transaction is RLP-encoded and can never begin with 0x19, so a\n" +
    "prefixed signature is structurally incapable of being replayed as a transfer\n" +
    "of your funds. This is EIP-191. Any site that asks you to sign unprefixed\n" +
    "raw bytes is asking for something it should not have.\n",
);

// ---------------------------------------------------------------------------
// Part 3 — break it on purpose: the signature never expires.
// ---------------------------------------------------------------------------

console.log("--- Part 3: the replay ---");

/** A naive login check: does this signature belong to this address? */
async function naiveLogin(address: Hex, sig: Hex): Promise<boolean> {
  return verifyMessage({ address, message: "Log me in to Metaspace", signature: sig });
}

console.log("first login  :", await naiveLogin(player.address, signature));
console.log("same sig, an hour later :", await naiveLogin(player.address, signature));
console.log("same sig, from a different IP, forever :", await naiveLogin(player.address, signature));

console.log(
  "\nThat is a fatal bug, and it is the most common one in hand-rolled wallet\n" +
    "login. The signature is a bearer token with no expiry. Anyone who sees it —\n" +
    "in a log file, in a proxy, in a leaked database — can log in as that player\n" +
    "for the rest of time, without ever touching the key.\n\n" +
    "The fix has three parts, and together they are Sign-In With Ethereum\n" +
    "(EIP-4361), which level 5 builds. The server issues a random nonce and\n" +
    "records it; the message the player signs contains that nonce, the domain of\n" +
    "the site and an expiry; and the server deletes the nonce the moment it is\n" +
    "used, so the second presentation of the same signature finds nothing to\n" +
    "match. Notice that all three are server-side state. The wallet cannot make\n" +
    "a login secure on its own.",
);

// ---------------------------------------------------------------------------
// Part 4 — the other replay: across chains and contracts.
// ---------------------------------------------------------------------------

console.log("\n--- Part 4: the replay you cannot see ---");

const voucherish = "mint item 42 to 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
const voucherSig = await player.signMessage({ message: voucherish });

console.log("a 'voucher' signed as a plain string:", voucherish);
console.log("signature:", voucherSig);
console.log(
  "\nThis signature is valid on Polygon, on Ethereum, on every testnet, and to\n" +
    "every contract that happens to check the same string — because nothing in\n" +
    "what was signed says which chain or which contract it was meant for. A game\n" +
    "that authorised mints this way could have its testnet vouchers replayed on\n" +
    "mainnet.\n\n" +
    "EIP-712 fixes this by hashing a typed struct together with a domain\n" +
    "separator that pins the contract name, version, chainId and verifying\n" +
    "address. That is level 10, and it is the single most interview-relevant\n" +
    "thing in this whole ladder for a game studio that mints assets.",
);

File: metaspace/output/exercises/python/04_sign_and_recover.py
"""
Level 4, in Python — the counterpart to solutions/04_sign_and_recover.ts.

Signing and recovery are where Python is at its strongest in this stack:
eth-account is mature, the API is small, and nothing about it is awkward.

Run: ./.venv/bin/python 04_sign_and_recover.py     (no network, no funds)
"""

from eth_account import Account
from eth_account.messages import encode_defunct, encode_typed_data
from eth_utils import keccak

# Anvil account #0: published in every toolchain's docs, holds nothing.
PLAYER_KEY = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
player = Account.from_key(PLAYER_KEY)

# --- Part 1: sign, then recover -------------------------------------------

message = "Log me in to Metaspace"
signable = encode_defunct(text=message)          # this applies the EIP-191 prefix
signed = Account.sign_message(signable, private_key=player.key)

print("--- Part 1: the round trip ---")
print("signer address :", player.address)
print("signature      :", signed.signature.hex()[:24] + "…")
print("  r :", hex(signed.r)[:20] + "…")
print("  s :", hex(signed.s)[:20] + "…")
print("  v :", signed.v)

recovered = Account.recover_message(signable, signature=signed.signature)
print("recovered      :", recovered)
print("matches signer :", recovered == player.address)

# --- Part 2: the prefix, by hand ------------------------------------------

prefix = f"\x19Ethereum Signed Message:\n{len(message)}".encode()
by_hand = keccak(prefix + message.encode())

print("\n--- Part 2: what actually got signed ---")
print("hash by hand   :", by_hand.hex())
print("eth-account's  :", signable.body.hex() if hasattr(signable, "body") else "n/a")
print(
    "\nSame construction as the TypeScript: the key never signs your message, it\n"
    "signs a hash of the message with a fixed prefix glued on, and that prefix is\n"
    "what makes the signature structurally unable to be a transaction (EIP-191)."
)

# --- Part 3: the replay ---------------------------------------------------

print("\n--- Part 3: the replay ---")
for attempt in ("first login", "an hour later", "from anywhere, forever"):
    ok = Account.recover_message(signable, signature=signed.signature) == player.address
    print(f"  {attempt:<26} -> {ok}")

print(
    "\nIdentical to the TypeScript, and identically broken: a bare signature is a\n"
    "bearer token with no expiry. The fix is server-side state — a nonce issued,\n"
    "recorded and burned in one statement — not anything the wallet can do.\n"
)

# --- Part 4: EIP-712, the voucher -----------------------------------------

print("--- Part 4: typed data ---")

typed = {
    "types": {
        "EIP712Domain": [
            {"name": "name", "type": "string"},
            {"name": "version", "type": "string"},
            {"name": "chainId", "type": "uint256"},
            {"name": "verifyingContract", "type": "address"},
        ],
        "Voucher": [
            {"name": "to", "type": "address"},
            {"name": "itemId", "type": "uint256"},
            {"name": "amount", "type": "uint256"},
            {"name": "voucherId", "type": "uint256"},
            {"name": "deadline", "type": "uint256"},
        ],
    },
    "primaryType": "Voucher",
    "domain": {
        "name": "Metaspace Items",
        "version": "1",
        "chainId": 80002,
        "verifyingContract": "0x5FbDB2315678afecb367f032d93F642f64180aa3",
    },
    "message": {
        "to": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
        "itemId": 42,
        "amount": 1,
        "voucherId": 9001,
        "deadline": 4102444800,
    },
}

signable_typed = encode_typed_data(full_message=typed)
signed_typed = Account.sign_message(signable_typed, private_key=player.key)
recovered_typed = Account.recover_message(signable_typed, signature=signed_typed.signature)

type_hash = keccak(
    b"Voucher(address to,uint256 itemId,uint256 amount,uint256 voucherId,uint256 deadline)"
)

print("type hash      :", type_hash.hex())
print("signature      :", signed_typed.signature.hex()[:24] + "…")
print("recovered      :", recovered_typed == player.address)
print(
    "\nThat type hash is the SAME 32 bytes the TypeScript computes and the same\n"
    "bytes solc embedded in the contract — run `npm run l10` and compare. Which\n"
    "is the real headline of this whole file: the interoperability is at the\n"
    "protocol, not the library. A Python backend can sign vouchers that a\n"
    "Solidity contract written against a TypeScript reference implementation\n"
    "accepts, because all three are hashing the same bytes."
)
